[Refactor Repositories] FileSystemRepository base class

joaoandre-avaiga opened the pull request:
So the objective was to keep the class as simple as possible, just having the responsability of writing and reading data from the database backend, in this case the File System. I've kept most of the version system logic that we have on the current version, but that can be also revisitated on future tasks. Some things I've removed maybe need to be kept, like the retry parameter that was on the old __to_entity method, which has been changed in this PR. Let me know if retry is still needed, I don't think we do in the other repositories.

_to_entity and _from_entity have been moved to the respective model class, I think it makes sense to use the model as the bridge between the taipy objects and the data in the database. Of course this can be changed. You can treat this PR as a proposal, and we can and should continue to propose better changes as we go futher into the repository refactoring.

joaoandre-avaiga committed:
feat: Base FileSystem repository refactor

joaoandre-avaiga committed:
chore: add missing methods

joaoandre-avaiga committed:
chore: add missing methods

joaoandre-avaiga committed:
chore: add default parameter for version number

joaoandre-avaiga commented:
I've just realized that PRs to the dev/3.0 branch don't trigger GH Actions. I will add the branch to the GH actions while we work on this milestone if everyone agrees.

joaoandre-avaiga committed:
chore: add PRs to dev/3.0 in linter and tests workflows

toan-quach commented on src/taipy/core/common/typing.py:
```diff
@@ -0,0 +1,16 @@
+from typing import TypeVar, Union
+
+ModelType = TypeVar("ModelType")
+Entity = TypeVar("Entity")
```
there's a class called `_Entity` in file `src/taipy/core/common/_entity.py` that all entities extended, I think it will come to be handy for this :D

joaoandre-avaiga replied:
Yes, it was intended to remove this functions elsewhere. We can see where as we go alone. The idea is to decouple the base class to maintain only the responsability of communicating with the database itself.

toan-quach commented on src/taipy/core/_repository/_v2/_filesystem_repository.py:
```diff
@@ -0,0 +1,143 @@
+        r = []
+        try:
+            for f in self.dir_path.iterdir():
+                if entity := self.model._to_entity(self.model(**self._get(f))):
```
this can be a huge hit to the performance if we try to load all files. I think Florian did a very neat trick in the function `__match_file_and_get_entity` the current version of file `src/taipy/core/_repository/_filesystem_repository.py` where you read the file as a text file, look for the "data" in the form of text, if it exists in the file, read it as an entity. It still requires us to load all files but not as a taipy entity but as texts, do you think this will work in this case?

joaoandre-avaiga replied:
I've added the logic of first loading the file contents and filtering before converting to a entity. I've remove some of logic that handled version numbers, but other than that it should behave very closely to the old repository. Of course we can always test after the FSRepo is done.

toan-quach commented on src/taipy/core/_repository/_v2/_filesystem_repository.py:
```diff
@@ -0,0 +1,143 @@
+            pass
+        return r
+
+    def _load_all_by(self, by, version_number: Optional[str]) -> List[Entity]:
```
this looks the same as `load_all` above, is there a reason for it?

joaoandre-avaiga replied:
Yes, I agree, theres no reason to have both. I've added an entra optional parameter in load so we can have filters, it's a dictionary that we can use to search in the document.

toan-quach reviewed (approved):
I noticed some of the function that was available in the old FSrepository class that is not implemented here. I had a look and they are not used anywhere which is weird cause if I remember correctly, they are used for optimizing the performance of entity creation in FS repos. Some of the function names are: `_get_by_config_and_owner_id`, `_get_by_configs_and_owner_ids` and `__match_file_and_get_entity`. 

Other than that, everything looks good to me! Just have some minor remarks 😃

jrobinAV commented on src/taipy/core/_repository/_v2/_filesystem_repository.py:
```diff
@@ -0,0 +1,143 @@
+from taipy.config.config import Config
+
+from ...exceptions import InvalidExportPath, ModelNotFound
+from ._repository import _AbstractRepository, _CustomDecoder, _CustomEncoder
```
```suggestion
from ._repository.v2 import _AbstractRepository, _CustomDecoder, _CustomEncoder
```

jrobinAV reviewed (commented):
- [x] In the AbstractRepository, the class name and the module name should be aligned
- [x] In the AbstractRepository, `load` method should start with an '_'.
- [x] In the AbstractRepository, _Encoder and _Decoder don't need to be prefixed by custom and should be in dedicated modules.
- [x] In the AbstractRepository, the version_number should be replaced by a generic field/value parameter.

- [x] FSRepository should inherit from the _AbstractRepository v2 class
- [x] In FSRepository, The version_number should be replaced by a generic field/value pair. It should remain the responsibility of the EntityFSRepository and not the FSrepository.
- [x] In FSRepository, the retry pattern must be re-implemented. Since the FS repo is not transactional, it aims to avoid conflicts in accessing the file.

joaoandre-avaiga committed:
chore: update naming in AbstractRepository

joaoandre-avaiga committed:
feat: add retry decorator

joaoandre-avaiga commented:
Im not sure I understood the point about the version number, maybe we can talk about it on the daily tomorrow

jrobinAV committed:
Merge branch 'dev/3.0' into feature/core#495

joaoandre-avaiga committed:
feat: load text before converting to object when searching for document

joaoandre-avaiga commented:
> I noticed some of the function that was available in the old FSrepository class that is not implemented here. I had a look and they are not used anywhere which is weird cause if I remember correctly, they are used for optimizing the performance of entity creation in FS repos. Some of the function names are: `_get_by_config_and_owner_id`, `_get_by_configs_and_owner_ids` and `__match_file_and_get_entity`.
> 
> Other than that, everything looks good to me! Just have some minor remarks smiley

This is by design. I think we should be able to have a similar behavior with the current base methods, the new `_load` with the filters parameters for instance. This would be an use case to test when we move to the managers implementation, which is the place the calls this methods today.

toan-quach commented on src/taipy/core/_repository/_v2/_abstract_repository.py:
```diff
@@ -0,0 +1,167 @@
+
+Json = Union[dict, list, str, int, float, bool, None]
+
+ModelType = TypeVar("ModelType")
+Entity = TypeVar("Entity")
```
@jrobinAV @joaoandre-avaiga 
I noticed we have the use of TypeVar for ModelType and EntityType in many places (_filesystem_repository.py, _manager.py, _sql_repository.py). Does it make sense to define and share them all in 1 place (alias.py perhaps)?

joaoandre-avaiga replied:
Yeah I've moved them to `taipy.core.common.typing`, just forgot to import it on this file. Thanks!

joaoandre-avaiga replied:
I had to rollback this change because mypy is no reconizing TypeVar defined in another file as a generic type, so this breaks the linter.

jrobinAV commented on tests/core/repository/test_fs_repository_v2.py:
```diff
@@ -0,0 +1,151 @@
+        r._export("uuid", export_path)
+        assert pathlib.Path(os.path.join(export_path, "mock_model/uuid.json")).exists()
+        with open(os.path.join(export_path, "mock_model/uuid.json"), "r") as exported_file:
+            exported_data = json._load(exported_file)
```
```suggestion
            exported_data = json.load(exported_file)
```

joaoandre-avaiga replied:
I did a mass replace when we changed the name from `load` to `_load`, this is a mistake.

jrobinAV commented on src/taipy/core/_repository/_v2/_filesystem_repository.py:
```diff
@@ -0,0 +1,135 @@
+        for model_id in ids:
+            self._delete(model_id)
+
+    def _search(self, attribute: str, value: Any, filters: List[Dict] = None) -> Optional[Entity]:
```
The signature is different from the parent class _AbstractRepository. Is it on purpose?

joaoandre-avaiga replied:
Fixed!

joaoandre-avaiga committed:
chore: import types instead of redefining them

joaoandre-avaiga commented:
> * [x]  In the AbstractRepository, the class name and the module name should be aligned
> * [x]  In the AbstractRepository, `load` method should start with an '_'.
> * [x]  In the AbstractRepository, _Encoder and _Decoder don't need to be prefixed by custom and should be in dedicated modules.
> * [x]  In the AbstractRepository, the version_number should be replaced by a generic field/value parameter.
> * [x]  FSRepository should inherit from the _AbstractRepository v2 class
> * [x]  In FSRepository, The version_number should be replaced by a generic field/value pair. It should remain the responsibility of the EntityFSRepository and not the FSrepository.
> * [x]  In FSRepository, the retry pattern must be re-implemented. Since the FS repo is not transactional, it aims to avoid conflicts in accessing the file.

This should be ok now. I've added the retry logic back with a decorator and removed all version references from the base class. Im thinking of implementing a `VersionMixin` to handle the version number and passing the version using the new `filters` parameter on the _load_all method.

joaoandre-avaiga committed:
fix: mypy does not reconize typevar defined in another file

joaoandre-avaiga committed:
chore: remove type ignore

jrobinAV commented on src/taipy/core/_repository/_v2/_abstract_repository.py:
```diff
@@ -15,15 +15,12 @@
+from typing import Any, Dict, Generic, Iterable, List, Optional, Union
 
-Json = Union[dict, list, str, int, float, bool, None]
+from src.taipy.core.common.typing import Entity, Json, ModelType
```
I believe it must be a relative path. otherwise it won't work on taipy/enterprise/rest repos.

jrobinAV commented on src/taipy/core/_repository/_v2/_abstract_repository.py:
```diff
@@ -15,9 +15,12 @@
+from typing import Any, Dict, Generic, Iterable, List, Optional, TypeVar, Union
 
-from src.taipy.core.common.typing import Entity, Json, ModelType
+from src.taipy.core.common.typing import Json
```
I believe it must be a relative path

jrobinAV commented on src/taipy/core/_repository/_v2/_abstract_repository.py:
```diff
@@ -23,7 +23,7 @@
 
 
-class _AbstractRepository(Generic[ModelType, Entity]):  # type: ignore
+class _AbstractRepository(Generic[ModelType, Entity]):
```
Yeah !!! we removed a type ignore. We are more often prone to add new ones...

joaoandre-avaiga replied:
This one I had added as well :smile:, but now it's gone.

jrobinAV replied:
👍🏻

joaoandre-avaiga committed:
fix: relative import
